File: sdk/ios/Sources/ExperimentationSDK/Cache.swift
import Foundation

// MARK: - CacheBox

/// A reference-type wrapper that stores a value alongside its expiry date.
/// Required because NSCache only accepts AnyObject as its value type.
private final class CacheBox<T> {
    let value: T
    let expiresAt: Date

    init(_ value: T, expiresAt: Date) {
        self.value = value
        self.expiresAt = expiresAt
    }

    var isExpired: Bool { Date() > expiresAt }
}

// MARK: - Cache

/// A thread-safe, size-bounded, TTL-aware cache backed by NSCache.
///
/// Key must be an `AnyObject` subclass (e.g., `NSString`).
/// Value can be any Swift type; it is wrapped internally in a `CacheBox`.
public final class Cache<Key: AnyObject, Value: AnyObject> {
    private let nsCache: NSCache<Key, CacheBox<Value>>
    private let ttl: TimeInterval
    private let lock = NSLock()

    /// - Parameters:
    ///   - maxSize: Maximum number of entries. NSCache may evict earlier under memory pressure.
    ///   - ttl: Time-to-live for each entry in seconds.
    public init(maxSize: Int, ttl: TimeInterval) {
        self.ttl = ttl
        self.nsCache = NSCache<Key, CacheBox<Value>>()
        self.nsCache.countLimit = maxSize
    }

    /// Returns the cached value for `key`, or `nil` if it is absent or expired.
    public func get(_ key: Key) -> Value? {
        lock.lock()
        defer { lock.unlock() }
        guard let box = nsCache.object(forKey: key) else { return nil }
        if box.isExpired {
            nsCache.removeObject(forKey: key)
            return nil
        }
        return box.value
    }

    /// Stores `value` under `key` with a TTL expiry.
    public func set(_ key: Key, value: Value) {
        lock.lock()
        defer { lock.unlock() }
        let box = CacheBox(value, expiresAt: Date().addingTimeInterval(ttl))
        nsCache.setObject(box, forKey: key)
    }

    /// Removes the entry for `key`.
    public func remove(_ key: Key) {
        lock.lock()
        defer { lock.unlock() }
        nsCache.removeObject(forKey: key)
    }

    /// Removes all cached entries.
    public func removeAll() {
        lock.lock()
        defer { lock.unlock() }
        nsCache.removeAllObjects()
    }
}

// MARK: - UserKeyCache

/// A thread-safe, TTL-aware cache of server results keyed by **user + key**, backed by NSCache.
///
/// Used by ``ExperimentationClient`` for flag evaluations and experiment assignments. Only
/// successful results are ever stored. Unlike a bare NSCache it can enumerate the live entries
/// of one user (needed for the track fan-out), via a small per-user key index.
public final class UserKeyCache<Value> {
    private let store = NSCache<NSString, CacheBox<Value>>()
    /// userId → keys in insertion order. Entries evicted by NSCache are dropped lazily.
    private var index: [String: [String]] = [:]
    private let lock = NSLock()
    private let ttl: TimeInterval

    /// - Parameters:
    ///   - ttl: Time-to-live for each entry in seconds.
    ///   - countLimit: Advisory maximum number of entries (NSCache `countLimit`).
    public init(ttl: TimeInterval, countLimit: Int = 1000) {
        self.ttl = ttl
        store.countLimit = countLimit
    }

    /// Returns the cached value for the user + key, or `nil` if absent or expired.
    public func get(userId: String, key: String) -> Value? {
        lock.lock()
        defer { lock.unlock() }
        let storageKey = Self.storageKey(userId, key)
        guard let box = store.object(forKey: storageKey) else {
            dropFromIndex(userId, key)
            return nil
        }
        if box.isExpired {
            store.removeObject(forKey: storageKey)
            dropFromIndex(userId, key)
            return nil
        }
        return box.value
    }

    /// Stores `value` for the user + key with a TTL expiry.
    public func set(userId: String, key: String, value: Value) {
        lock.lock()
        defer { lock.unlock() }
        let box = CacheBox(value, expiresAt: Date().addingTimeInterval(ttl))
        store.setObject(box, forKey: Self.storageKey(userId, key))
        var keys = index[userId] ?? []
        if !keys.contains(key) {
            keys.append(key)
        }
        index[userId] = keys
    }

    /// All live (non-expired) entries for a user, in insertion order.
    public func entries(userId: String) -> [(key: String, value: Value)] {
        lock.lock()
        defer { lock.unlock() }
        var live: [(key: String, value: Value)] = []
        var liveKeys: [String] = []
        for key in index[userId] ?? [] {
            let storageKey = Self.storageKey(userId, key)
            if let box = store.object(forKey: storageKey), !box.isExpired {
                live.append((key: key, value: box.value))
                liveKeys.append(key)
            } else {
                store.removeObject(forKey: storageKey)
            }
        }
        index[userId] = liveKeys.isEmpty ? nil : liveKeys
        return live
    }

    /// Removes the entry for the user + key.
    public func remove(userId: String, key: String) {
        lock.lock()
        defer { lock.unlock() }
        store.removeObject(forKey: Self.storageKey(userId, key))
        dropFromIndex(userId, key)
    }

    /// Removes all cached entries.
    public func removeAll() {
        lock.lock()
        defer { lock.unlock() }
        store.removeAllObjects()
        index.removeAll()
    }

    /// Number of live entries across all users.
    public var count: Int {
        lock.lock()
        let users = Array(index.keys)
        lock.unlock()
        return users.reduce(0) { $0 + entries(userId: $1).count }
    }

    // MARK: Private

    /// Length-prefixed so a user ID containing the separator cannot collide with another pair.
    private static func storageKey(_ userId: String, _ key: String) -> NSString {
        "\(userId.utf8.count):\(userId):\(key)" as NSString
    }

    private func dropFromIndex(_ userId: String, _ key: String) {
        guard var keys = index[userId] else { return }
        keys.removeAll { $0 == key }
        index[userId] = keys.isEmpty ? nil : keys
    }
}

File: sdk/dotnet/src/ExperimentationPlatform/Cache.cs
namespace ExperimentationPlatform;

/// <summary>
/// Thread-safe in-memory cache with TTL expiration and LRU-style eviction.
/// </summary>
/// <typeparam name="TValue">The type of cached values.</typeparam>
public class SdkCache<TValue>
{
    private readonly int _maxSize;
    private readonly TimeSpan _defaultTtl;

    // Dictionary<K,V> does NOT guarantee enumeration order: after a removal the
    // freed slot is reused by the next insertion, so "the first key enumerated"
    // is not the oldest and eviction would drop an arbitrary entry. Insertion
    // order is tracked explicitly instead.
    private readonly Dictionary<string, CacheEntry> _store;
    private readonly LinkedList<string> _insertionOrder = new();
    private readonly Dictionary<string, LinkedListNode<string>> _orderNodes = new();
    private readonly object _lock = new();

    private struct CacheEntry
    {
        public TValue Value;
        public DateTime ExpiresAt;
    }

    /// <summary>
    /// Creates a new cache.
    /// </summary>
    /// <param name="maxSize">Maximum number of entries before oldest is evicted.</param>
    /// <param name="defaultTtl">Default time-to-live for entries.</param>
    public SdkCache(int maxSize = 1000, TimeSpan? defaultTtl = null)
    {
        _maxSize = maxSize;
        _defaultTtl = defaultTtl ?? TimeSpan.FromSeconds(300);
        _store = new Dictionary<string, CacheEntry>(_maxSize);
    }

    /// <summary>Gets a cached value, or default if not found or expired.</summary>
    /// <remarks>
    /// For a value type this cannot distinguish "missing" from a cached default
    /// (<c>SdkCache&lt;int&gt;.Get</c> returns 0 either way). Use
    /// <see cref="TryGet"/> when that difference matters.
    /// </remarks>
    public TValue? Get(string key)
    {
        return TryGet(key, out var value) ? value : default;
    }

    /// <summary>Gets a cached value; false when the key is absent or expired.</summary>
    public bool TryGet(string key, out TValue value)
    {
        lock (_lock)
        {
            if (!_store.TryGetValue(key, out var entry))
            {
                value = default!;
                return false;
            }

            if (DateTime.UtcNow >= entry.ExpiresAt)
            {
                RemoveLocked(key);
                value = default!;
                return false;
            }

            value = entry.Value;
            return true;
        }
    }

    /// <summary>Removes a key and its order entry. Caller holds the lock.</summary>
    private void RemoveLocked(string key)
    {
        _store.Remove(key);
        if (_orderNodes.TryGetValue(key, out var node))
        {
            _insertionOrder.Remove(node);
            _orderNodes.Remove(key);
        }
    }

    /// <summary>Sets a value in the cache.</summary>
    /// <param name="key">The cache key.</param>
    /// <param name="value">The value to store.</param>
    /// <param name="ttl">Override TTL. Uses the cache's default TTL if null.</param>
    public void Set(string key, TValue value, TimeSpan? ttl = null)
    {
        lock (_lock)
        {
            // If key already exists, remove it first to refresh insertion order.
            RemoveLocked(key);

            // Evict the oldest entry if at capacity.
            while (_store.Count >= _maxSize && _insertionOrder.First is { } oldest)
            {
                RemoveLocked(oldest.Value);
            }

            _store[key] = new CacheEntry
            {
                Value = value,
                ExpiresAt = DateTime.UtcNow + (ttl ?? _defaultTtl)
            };
            _orderNodes[key] = _insertionOrder.AddLast(key);
        }
    }

    /// <summary>Removes a specific key from the cache.</summary>
    public void Delete(string key)
    {
        lock (_lock)
        {
            RemoveLocked(key);
        }
    }

    /// <summary>Removes all entries from the cache.</summary>
    public void Clear()
    {
        lock (_lock)
        {
            _store.Clear();
            _insertionOrder.Clear();
            _orderNodes.Clear();
        }
    }

    /// <summary>Returns the current number of entries in the cache (including expired ones not yet purged).</summary>
    public int Count
    {
        get
        {
            lock (_lock)
            {
                return _store.Count;
            }
        }
    }

    /// <summary>
    /// Returns the number of non-expired entries in the cache.
    /// Note: This iterates through all entries and is O(n).
    /// </summary>
    public int ActiveCount
    {
        get
        {
            lock (_lock)
            {
                var now = DateTime.UtcNow;
                int count = 0;
                foreach (var entry in _store.Values)
                {
                    if (now < entry.ExpiresAt)
                        count++;
                }
                return count;
            }
        }
    }
}

/// <summary>
/// Thread-safe, TTL-aware cache of server results keyed by <b>user + key</b>, built on
/// <see cref="SdkCache{TValue}"/>. Used by <see cref="ExperimentationClient"/> for flag
/// evaluations and experiment assignments; only successful results are ever stored. Unlike a bare
/// <see cref="SdkCache{TValue}"/> it can enumerate the live entries of one user (needed for the
/// track fan-out) via a small per-user key index.
/// </summary>
/// <typeparam name="TValue">The cached value type.</typeparam>
public class UserKeyCache<TValue> where TValue : class
{
    private readonly SdkCache<TValue> _store;
    // userId → keys in insertion order. Entries evicted or expired in _store are dropped lazily.
    private readonly Dictionary<string, List<string>> _index = new();
    private readonly object _lock = new();

    /// <summary>Creates a new cache.</summary>
    /// <param name="maxSize">Maximum number of entries across all users before the oldest is evicted.</param>
    /// <param name="ttl">Time-to-live of each entry.</param>
    public UserKeyCache(int maxSize, TimeSpan ttl)
    {
        _store = new SdkCache<TValue>(maxSize, ttl);
    }

    /// <summary>Returns the cached value for the user + key, or <c>null</c> if absent or expired.</summary>
    public TValue? Get(string userId, string key)
    {
        lock (_lock)
        {
            var value = _store.Get(StorageKey(userId, key));
            if (value == null)
                DropFromIndex(userId, key);
            return value;
        }
    }

    /// <summary>Stores <paramref name="value"/> for the user + key with the cache TTL.</summary>
    public void Set(string userId, string key, TValue value)
    {
        lock (_lock)
        {
            _store.Set(StorageKey(userId, key), value);
            if (!_index.TryGetValue(userId, out var keys))
            {
                keys = new List<string>();
                _index[userId] = keys;
            }
            if (!keys.Contains(key))
                keys.Add(key);
        }
    }

    /// <summary>All live (non-expired) entries of a user, in insertion order.</summary>
    public IReadOnlyList<KeyValuePair<string, TValue>> Entries(string userId)
    {
        lock (_lock)
        {
            var live = new List<KeyValuePair<string, TValue>>();
            if (!_index.TryGetValue(userId, out var keys))
                return live;

            var liveKeys = new List<string>();
            foreach (var key in keys)
            {
                var value = _store.Get(StorageKey(userId, key));
                if (value != null)
                {
                    live.Add(new KeyValuePair<string, TValue>(key, value));
                    liveKeys.Add(key);
                }
            }

            if (liveKeys.Count == 0)
                _index.Remove(userId);
            else
                _index[userId] = liveKeys;
            return live;
        }
    }

    /// <summary>Removes the entry for the user + key.</summary>
    public void Remove(string userId, string key)
    {
        lock (_lock)
        {
            _store.Delete(StorageKey(userId, key));
            DropFromIndex(userId, key);
        }
    }

    /// <summary>Removes all entries.</summary>
    public void Clear()
    {
        lock (_lock)
        {
            _store.Clear();
            _index.Clear();
        }
    }

    /// <summary>Number of live entries across all users.</summary>
    public int Count
    {
        get
        {
            List<string> users;
            lock (_lock)
            {
                users = new List<string>(_index.Keys);
            }
            int total = 0;
            foreach (var userId in users)
                total += Entries(userId).Count;
            return total;
        }
    }

    // Length-prefixed so a user ID containing the separator cannot collide with another pair.
    private static string StorageKey(string userId, string key) => $"{userId.Length}:{userId}:{key}";

    private void DropFromIndex(string userId, string key)
    {
        if (!_index.TryGetValue(userId, out var keys))
            return;
        keys.Remove(key);
        if (keys.Count == 0)
            _index.Remove(userId);
    }
}
